### PR TITLE
Bug 1363449 - Include MMA / Focus work in Firefox for iOS 7.5

### PR DESCRIPTION
**Pull request by Sdaswani**

### CODE REVIEW

**Comment by farhanpatel** [comment]
### SwiftLint found issues

#### Warnings

File | Line | Reason |
| --- | ----- | ----- |
AppDelegate.swift | 142 | When registing for a notification using a block, the opaque observer that is returned should be stored so it can be removed later. 
AppDelegate.swift | 149 | When registing for a notification using a block, the opaque observer that is returned should be stored so it can be removed later. 
BrowserViewController.swift | 995 | Prefer `!= nil` over `let _ =` 
BrowserViewController.swift | 1416 | Prefer `!= nil` over `let _ =` 
BrowserViewController.swift | 1706 | Prefer `!= nil` over `let _ =` 
BrowserViewController.swift | 1961 | Prefer `!= nil` over `let _ =` 
BrowserViewController.swift | 2449 | Prefer `!= nil` over `let _ =` 
BrowserViewController.swift | 3224 | Prefer `!= nil` over `let _ =` 
BrowserViewController.swift | 1873 | `where` clauses are preferred over a single `if` inside a `for`. 
BrowserViewController.swift | 1057 | Prefer `_ = foo()` over `let _ = foo()` when discarding a result from a function. 
BrowserViewController.swift | 1074 | Prefer `_ = foo()` over `let _ = foo()` when discarding a result from a function. 
BrowserViewController.swift | 2891 | Prefer `_ = foo()` over `let _ = foo()` when discarding a result from a function. 
LoginsHelper.swift | 98 | When the index is not used, `.enumerated()` can be removed. 
TabManager.swift | 720 | When the index is not used, `.enumerated()` can be removed. 
TabManager.swift | 422 | Prefer `!= nil` over `let _ =` 
TabManager.swift | 152 | `where` clauses are preferred over a single `if` inside a `for`. 
TabManager.swift | 164 | `where` clauses are preferred over a single `if` inside a `for`. 
TabManager.swift | 529 | `where` clauses are preferred over a single `if` inside a `for`. 
TabManager.swift | 782 | Prefer `_ = foo()` over `let _ = foo()` when discarding a result from a function. 
TabTrayController.swift | 620 | Prefer `!= nil` over `let _ =` 
TabTrayController.swift | 855 | `where` clauses are preferred over a single `if` inside a `for`. 
TabTrayController.swift | 639 | Prefer `_ = foo()` over `let _ = foo()` when discarding a result from a function. 
TabTrayController.swift | 683 | Prefer `_ = foo()` over `let _ = foo()` when discarding a result from a function. 
TabTrayController.swift | 718 | Prefer `_ = foo()` over `let _ = foo()` when discarding a result from a function. 
TabTrayController.swift | 1116 | Prefer `_ = foo()` over `let _ = foo()` when discarding a result from a function. 
BookmarksPanel.swift | 392 | Prefer `_ = foo()` over `let _ = foo()` when discarding a result from a function. 
AppSettingsOptions.swift | 175 | Prefer `!= nil` over `let _ =` 
AppSettingsOptions.swift | 352 | Prefer `!= nil` over `let _ =` 


<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

File: bootstrap.sh [hunk 1]
```diff
@@ -29,6 +29,6 @@ if [ ! -z "$XCS_BOT_ID"  ]; then
 fi
 
 if [ ! -f Carthage/Cartfile.resolved ]; then
-  carthage bootstrap $CARTHAGE_VERBOSE --platform ios --color auto --no-use-binaries
+  carthage bootstrap $CARTHAGE_VERBOSE --platform ios --color auto
```
Comment (jhugman) [inline bootstrap.sh]: Put this in another PR please.

File: Client/Frontend/Browser/TabManager.swift [hunk 2]
```diff
@@ -453,14 +453,40 @@ class TabManager: NSObject {
...
         delegates.forEach { $0.get()?.tabManagerDidRemoveAllTabs(self, toast: toast) }
     }
-    
+
+    func showFocusPromoToast() {
+        if !LeanplumIntegration.sharedInstance.shouldShowFocusUI() {
+            return
+        }
+
+        guard let templateDict = LeanplumIntegration.sharedInstance.getTemplateDictionary(), let labelText = templateDict["Template Text"], let buttonText = templateDict["Button Text"], let colorText = templateDict["Hex Color String"], let deepLink = templateDict["Deep Link"] else {
```
Comment (jhugman) [inline Client/Frontend/Browser/TabManager.swift]: Nit: put each of the `let`s on different lines.

File: Client/Frontend/Browser/TabManager.swift [hunk 3]
```diff
@@ -453,14 +453,40 @@ class TabManager: NSObject {
...
+            return
+        }
+
+        LeanplumIntegration.sharedInstance.track(eventName: .focusPromoImpression, withParameters: ["button name": buttonText as AnyObject])
+        let toast = LeanplumPromoButtonToast(labelText: labelText, buttonText: buttonText, colorText: colorText, completion: { buttonPressed in
+            guard let url = URL(string: deepLink) else {
+                return
+            }
+
+            if UIApplication.shared.canOpenURL(url) && buttonPressed {
```
Comment (jhugman) [inline Client/Frontend/Browser/TabManager.swift]: These guards and `canOpenURL`: they should be done before line the `.track` call on L475: there is no need to show the button if `deepLink` isn't a valid link that the device can open.

File: bootstrap.sh (hunk 1 above)
Comment (st3fan) [inline bootstrap.sh]: I am pretty sure this was put in for a reason?

File: Client/Frontend/Browser/TabManager.swift (hunk 3 above)
Comment (st3fan) [inline Client/Frontend/Browser/TabManager.swift]: Lets take a quick look at this, I don't know what  the intend here is. Not sure if this is incorrect.

File: Client/Application/LeanplumIntegration.swift [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+        return nil
+    }
+
+    // Utils
+    
+    func setEnabled(_ enabled: Bool) {
+        self.enabled = enabled
+        if self.enabled {
+            self.start()
+        }
```
Comment (st3fan) [inline Client/Application/LeanplumIntegration.swift]: If I understand correctly, there is `start()` for app startup. And `LeanplumIntegration.sharedInstance.setEnabled()` for then the user flips the *Send Anonymous Usage Data* switch.

I think this is good. My only request here is that we make sure that flipping *Send Anonymous Usage Data* actually disables LeanPlum. And that it does not wait until the app is cold-booted again. (I think it does, but maybe we can verify and make sure that events are not tracked / sent as soon as we flip that switch?)

Comment (Sdaswani) [inline Client/Application/LeanplumIntegration.swift]: @jhugman unfortunately there is no LeanPlum.stop - I checked with them. @st3fan flipping off _Send Anonymous Usage Data_ does indeed disable all further calls to LeanPlum in the current run. But LeanPlum is technically still running until the next cold-boot (since there is no stop).

File: Client/Application/LeanplumIntegration.swift [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+        guard self.enabled else {
+            return
+        }
+        
+        guard SupportedLocales(rawValue: Locale.current.identifier) != nil else {
+            return
+        }
+
+        if Leanplum.hasStarted() {
+            Logger.browserLogger.error("LeanplumIntegration - Already initialized")
```
Comment (Sdaswani) [inline Client/Application/LeanplumIntegration.swift]: @bkmunar should be quick to fix.

File: Client/Application/LeanplumIntegration.swift [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+            Leanplum.setAppId(settings.appId, withDevelopmentKey: settings.key)
+        case .production:
+            Logger.browserLogger.info("LeanplumIntegration - Setting up for Production")
+            Leanplum.setAppId(settings.appId, withProductionKey: settings.key)
+        }
+        Leanplum.syncResourcesAsync(true)
+        setupTemplateDictionary()
+
+        var userAttributesDict = [AnyHashable: Any]()
+        userAttributesDict["Focus Installed"] = "false"
```
Comment (Sdaswani) [inline Client/Application/LeanplumIntegration.swift]: @bkmunar should be quick to fix.

File: Client/Application/LeanplumIntegration.swift (hunk 4 above)
Comment (Sdaswani) [inline Client/Application/LeanplumIntegration.swift]: @jhugman fair point. @bkmunar make start() fileprivate, and then in the AppDelegate, just call LeanplumIntegration.setEnabled(true).

File: Client/Application/LeanplumIntegration.swift [hunk 7]
```diff
@@ -0,0 +1,196 @@
...
+    
+    func setEnabled(_ enabled: Bool) {
+        self.enabled = enabled
+        if self.enabled {
+            self.start()
+        }
+    }
+
+    func shouldShowFocusUI() -> Bool {
+        guard let shouldShowFocusUI = LeanplumIntegration.sharedInstance.getBoolVariableFromServer(key: "shouldShowFocusUI"), let focus = URL(string: "firefox-focus://"), let klar = URL(string: "firefox-klar://"), !UIApplication.shared.canOpenURL(focus) && !UIApplication.shared.canOpenURL(klar) && shouldShowFocusUI else {
```
Comment (Sdaswani) [inline Client/Application/LeanplumIntegration.swift]: @bkmunar should be quick to fix.

File: Client/Application/LeanplumIntegration.swift [hunk 8]
```diff
@@ -0,0 +1,196 @@
...
+            return false
+        }
+        return true
+    }
+
+    // Private
+
+    private func getSettings() -> LeanplumSettings? {
+        let bundle = Bundle.main
+        guard let environmentString = bundle.object(forInfoDictionaryKey: LeanplumEnvironmentKey) as? String, let environment = LeanplumEnvironment(rawValue: environmentString), let appId = bundle.object(forInfoDictionaryKey: LeanplumAppIdKey) as? String, let key = bundle.object(forInfoDictionaryKey: LeanplumKeyKey) as? String else {
```
Comment (Sdaswani) [inline Client/Application/LeanplumIntegration.swift]: @bkmunar should be quick to fix.

File: Client/Frontend/Browser/TabTrayController.swift [hunk 9]
```diff
@@ -1170,6 +1176,10 @@ extension TabTrayController: MenuActionDelegate {
                 DispatchQueue.main.async {
                     self.closeTabsForCurrentTray()
                 }
+
+                if privateMode {
+                    LeanplumIntegration.sharedInstance.track(eventName: .closedPrivateTabs)
```
Comment (Sdaswani) [inline Client/Frontend/Browser/TabTrayController.swift]: Fair enough @jhugman - I'm meeting with Legal / Privacy to address this concern already, and note that product will make the ultimate decision on this. As we discussed, we are only tracking telemetry, and nothing about the contents of your private session.

File: Client/Application/AppDelegate.swift [hunk 10]
```diff
@@ -157,10 +157,17 @@ class AppDelegate: UIResponder, UIApplicationDelegate, UIViewControllerRestorati
         
         adjustIntegration = AdjustIntegration(profile: profile)
 
+        LeanplumIntegration.sharedInstance.setup(profile: profile)
```
Comment (Sdaswani) [inline Client/Application/AppDelegate.swift]: @bkmunar should be quick to fix.

File: Client/Application/AppDelegate.swift [hunk 11]
```diff
@@ -157,10 +157,17 @@ class AppDelegate: UIResponder, UIApplicationDelegate, UIViewControllerRestorati
...
+        LeanplumIntegration.sharedInstance.start()
+
         // We need to check if the app is a clean install to use for
         // preventing the What's New URL from appearing.
         if getProfile(application).prefs.intForKey(IntroViewControllerSeenProfileKey) == nil {
```
Comment (Sdaswani) [inline Client/Application/AppDelegate.swift]: Sorry @jhugman - isn't @bkmunar just doing what the Adjust initialization is doing right above?

File: bootstrap.sh (hunk 1 above)
Comment (Sdaswani) [inline bootstrap.sh]: Yes, the reason is we don't compile LP, but we just import the binary, and the _--no-use-binaries_ option specifically disallows importing the binary. This was a fix suggested by @sleroux .

File: Client/Frontend/Browser/TabManager.swift (hunk 2 above)
Comment (Sdaswani) [inline Client/Frontend/Browser/TabManager.swift]: @bkmunar should be quick to fix.

File: Client/Frontend/Browser/TabManager.swift (hunk 3 above)
Comment (Sdaswani) [inline Client/Frontend/Browser/TabManager.swift]: That is a great point @jhugman - if we send down a bad url from LP's dashboard, we shouldn't show the toast if the user can't even act on it. Of course we would test this so it shouldn't be a _practical_ concern, but James is right to prevent it technically.

File: Client/Application/LeanplumIntegration.swift (hunk 4 above)
Comment (st3fan) [inline Client/Application/LeanplumIntegration.swift]: I don't think this is good enough. There is a user expectation that when they turn off *Send Anonymous User Data* that at that point they are opted out from data collection. We make this happens for Telemetry and Adjust, so Leanplum still collecting and sending data back cannot happen after this toggle has flipped to No.

I understand this will take effect the next time the user starts the application, but we have to be strict here. On a modern phone it is possible that Firefox runs for days or even weeks without being actually restarted.

Is there anything in the LeanPlum SDK that we can call to pause or stop the data collection?

(I know we have `shouldSendToLP()` but LP also does its down session tracking outside of calls to `Leanplum.track()`)

File: Client/Application/LeanplumIntegration.swift [hunk 12]
```diff
@@ -0,0 +1,207 @@
...
+
+        guard let settings = getSettings() else {
+            log.error("LeanplumIntegration - Could not load settings from Info.plist")
+            return
+        }
+
+        switch settings.environment {
+        case .development:
+            log.info("LeanplumIntegration - Setting up for Development")
+            Leanplum.setDeviceId(ASIdentifierManager.shared().advertisingIdentifier.uuidString)
```
Comment (Sdaswani) [inline Client/Application/LeanplumIntegration.swift]: @bkmunar cc @st3fan this used the IDFA, right? we need to use the IDFV.

File: Client/Application/LeanplumIntegration.swift [hunk 13]
```diff
@@ -0,0 +1,207 @@
...
+            Leanplum.setAppId(settings.appId, withDevelopmentKey: settings.key)
+        case .production:
+            log.info("LeanplumIntegration - Setting up for Production")
+            Leanplum.setAppId(settings.appId, withProductionKey: settings.key)
```
Comment (Sdaswani) [inline Client/Application/LeanplumIntegration.swift]: @bkmunar cc @st3fan we need to use the IDFV here to setDeviceID.
